cli/azd/pkg/ext/hooks.go

```diff
@@ -0,0 +1,379 @@
+		if scriptConfig == nil {
+			continue
+		}
+
+		index := slices.IndexFunc(validHookNames, func(hookName string) bool {
+			return hookName == scriptName
+		})
+
+		if index > -1 {
+			// If the script config includes an OS specific configuration use that instead
+			if runtime.GOOS == "windows" && scriptConfig.Windows != nil {
+				scriptConfig = scriptConfig.Windows
+			} else if (runtime.GOOS == "linux" || runtime.GOOS == "darwin") && scriptConfig.Linux != nil {
+				scriptConfig = scriptConfig.Linux
+			}
+
+			scriptConfig.Name = scriptName
+			scriptConfig.Path = strings.ReplaceAll(scriptConfig.Path, "/", string(os.PathSeparator))
+			matchingScripts = append(matchingScripts, scriptConfig)
+		}
+	}
+
+	return matchingScripts
+}
+
+func (h *CommandHooks) getImplicitHooks(prefix HookType, validHookNames []string) []*ScriptConfig {
+	matchingScripts := []*ScriptConfig{}
+
+	hooksDir := filepath.Join(h.cwd, ".azure", "hooks")
+	files, err := os.ReadDir(hooksDir)
+	if err != nil {
+		return matchingScripts
+	}
+
+	// Find implicit / convention based hooks in the .azure/hooks directory
+	// Find `predeploy.sh` or similar matching a hook prefix & valid command name
+	for _, file := range files {
+		fileName := file.Name()
+		fileNameWithoutExt := strings.TrimSuffix(fileName, path.Ext(fileName))
+
+		index := slices.IndexFunc(validHookNames, func(hookName string) bool {
+			return !file.IsDir() && hookName == fileNameWithoutExt
+		})
+
+		if index > -1 {
+			scriptType, err := inferScriptTypeFromFilePath(fileName)
+			if err != nil {
+				log.Printf("Found script hook '%s', but type is not supported. %s\n", fileName, err.Error())
+				continue
+			}
+
+			relativePath, err := filepath.Rel(h.cwd, filepath.Join(hooksDir, fileName))
+			if err != nil {
+				// This err should never happen since we are only looking for files inside the specified cwd.
+				log.Printf("Found script hook '%s', but is outside of project folder. Error: %s\n", file.Name(), err.Error())
+				continue
+			}
+
+			scriptConfig := ScriptConfig{
+				Name:     fileNameWithoutExt,
+				Path:     relativePath,
+				Location: ScriptLocationPath,
+				Type:     scriptType,
+			}
+
+			matchingScripts = append(matchingScripts, &scriptConfig)
+		}
+	}
+
+	return matchingScripts
+}
+
+func (h *CommandHooks) execScript(ctx context.Context, scriptConfig *ScriptConfig) error {
+	// Delete any temporary inline scripts after execution
+	defer func() {
+		if scriptConfig.Location == ScriptLocationInline {
+			os.Remove(scriptConfig.Path)
+		}
+	}()
+
+	script, err := h.GetScript(scriptConfig)
+	if err != nil {
+		return err
+	}
+
+	formatter := h.console.GetFormatter()
+	consoleInteractive := formatter == nil || formatter.Kind() == output.NoneFormat
+	scriptInteractive := consoleInteractive && scriptConfig.Interactive
+
+	// When running in an interactive terminal broadcast a message to the dev to remind them that custom hooks are running.
+	if consoleInteractive {
```

Comment:
I think we should broadcast this even if the format is `JSON`. It will be outputted as a machine-readable message in `stderr`.